### ISSUE TITLE
Volume Mounts with `:z` and `:Z` are dangerous

### ISSUE THREAD

**Issue by ElvenSpellmaker** [docker/docker#30934]
Hey there,

While this isn't a bug per se, it's _really_ unexpected behaviour of a Docker run command and I believe it deserves some attention, and at the _very_ least some clarification in the manual and steps to try and resolve the problem.

======

Tl;dr, `:z` and `:Z` can cause the system to disallow logins by changing labels (until a reboot at least) without so much as a warning or a single line of output.

======

I run a CentOS 7.3 VM in Azure for testing out Docker boxes and making new ones (plus a few other bits and pieces)

Today I ran a `docker run` (something similar to `docker run -v $PWD:/var/www:z -it <sha256 here> php -v`) in my home directory root (not usually where I run it) by doing a `^R` and searching for a command.
Because `$PWD` resolved to the my home dir the `:z` relabelled my _entire_ home directory with Docker's shared label. I wondered why the command took _so_ long to run, but didn't think much of it at the time.
I walked away from my machine and (luckily) left my ssh session on. Later on I tried to log into the server from my phone and couldn't.

After running back upstairs and some frantic defensive actions over the system (looking at `authorized_keys` files for tampering, audit logs, checking fail2ban, etc.) I decided the system had not been hacked into and my keys weren't compromised.

Eventually after searching around I noticed that `sshd` was being denied `read` to the `authorized_keys` file by SELinux because of the label change!

After a `chcon -RFvv /home/username` I could log in again.

======

While it's obvious that those options were changing the labels on the _host_, I didn't realise until this point. The manual on [Volume Labels](https://docs.docker.com/engine/tutorials/dockervolumes/#/volume-labels) mentions about changing host's labels, but doesn't mention how dangerous this is and that if you can't log into the system a reboot or a relabel will be needed... (What if this was a system I couldn't / couldn't easily reboot?)

It's possible the manual also should explicitly list the two labels that are assigned for public and private mounts, so that people who don't want to use `:z` or `:Z` can manually label the files beforehand, which'd be safer and also can be put into the permanent policy of the machine.

Thanks.

**Comment by thaJeztah** [comment]
By bind-mounting a host directory you're explicitly giving a container full access to anything inside that directory (on the host). In your case, this includes making your private keys accessible to the containers process (and probably even browsable in a web browser)

We may have to call this out more explicitly


ping @mstanleyjones - this was one of the things I wanted to mention for your rewrite of the volumes/mounts section

**Comment by justincormack** [comment]
Ugh. I definitely don't think that bind mounts should support relabelling at all. And overall it just seems the wrong place, a label should be set on volume create, but I do not expect a _mount_ to ever change the contents of a volume or bind at all. I think this was a terrible design mistake.

I think we should add an selinux label option to `docker volume create` (at least the local driver, but encourage other drivers to support it?) and deprecate this option entirely...

cc @riyazdf @runcom

**Comment by runcom** [comment]
mmm not sure what's really wrong with relabeling binds :confused: @rhatdan wrote http://www.projectatomic.io/blog/2016/03/dwalsh_selinux_containers/ and warned people when relabeling binds. I believe a huge, fat warning would be ok.

**Comment by justincormack** [comment]
@runcom because a bind mount having side effects of changing your system security labels is unacceptable. As the article points out you can disable selinux for that container if you want to mount stuff which you would not otherwise be able to due to the labels.

**Comment by ElvenSpellmaker** [comment]
That article is imo contains the label names that should be clearly in the documentation so that people may set up labels properly and in a controlled fashion, maybe even using proper management (such as `semanage fcontext`).

I have never really seen a programme do `chcon` itself, usually the user is required to set up the correct types beforehand.

It's _really_ obvious what it was doing don't get me wrong, but _only_ after the event, at the time I kind of assumed the labels were being changed _inside_ the container, not outside and the small snippet in the manual isn't _that_ clear.

Plus a blog post is hardly official documentation, you can't expect users to have to go and find blog posts to find out information about Docker, it's just the wrong place to send users or to expect users to find out criticla information imo.

**Comment by runcom** [comment]
@ElvenSpellmaker yeah, I was indeed suggesting to enhance the docs. Or follow @justincormack's path of disabling this altogether for binds (hard maybe?)

**Comment by rhatdan** [comment]
The problem with mounting directories into a container and the user setting up the labels, is the user can not know what the MCS label will be for the container process.  You could have changed the labels on the the mounted directory with a call like `chcon -t container_file_t MNTPOINT`, but this would setup that directory to be r/w for ALL of your containers from an SELinux point of view.  The only way to get a private label would be for the user to manage his own MCS labels and guarantee himself that his system would never reuse the labels.



```
chcon -t container_file_t -l s0:c1,c2 /MNTPNT
docker run --security-opt label=level:s0:c1,c2 -v /MNTPNT:/var/www ...
```

Here is what the man page says about using :z and :Z.  How would you want the wording changed?

```   Labeling  systems like SELinux require that proper labels are placed on
       volume content mounted into a container. Without a label, the  security
       system  might  prevent  the processes running inside the container from
       using the content. By default, Docker does not change the labels set by
       the OS.

       To  change  a label in the container context, you can add either of two
       suffixes :z or :Z to the volume mount. These suffixes  tell  Docker  to
       relabel  file  objects on the shared volumes. The z option tells Docker
       that two containers share the  volume  content.  As  a  result,  Docker
       labels  the  content  with a shared content label. Shared volume labels
       allow all containers to read/write content.  The Z option tells  Docker
       to  label  the content with a private unshared label.  Only the current
       container can use a private volume.

```

**Comment by leifmadsen** [comment]
In order to maybe move this forward a bit, I'd suggest this WARNING be added to the contents of the documentation:

> WARNING Use of :z or :Z can have impacts on your system if used too broadly. For example, the use of these options against a users HOME directory or other top level system directory would result in the relabeling of those directories for use by Docker, and could cause issues such as being unable to SSH back into a system. When using the relabel options, it should only be used on directories that are intended for use by containers, and not the host system itself. If your container does require broader access to system directories, then use of '--security-opt label:disable' with the 'docker run' command is a better alternative. Note that using the above option instead will disable SELinux checks for that container.

**Comment by rhatdan** [comment]
LGTM

**Comment by justincormack** [comment]
Created a docs PR https://github.com/docker/cli/pull/832

**Comment by abitrolly** [comment]
Why Docker needs to **re**label files? Can it just add labels for its own access without touching existing SELinux labels?

**Comment by rhatdan** [comment]
No.  The container runtimes pick a unigue SELinux labels to run for their processes and writable content needs to match.  This means that the user can not pre-know the labels.  By passing in the :z or :Z the user is telling the container runtime to change the labels to something that will work and be secure within the container.  Sadly if the user does not understand what he is doing this can lead to problems with confined domains on his host.  We try to block some of the common erros like relaying /var, /usr, /.  But we can not block all.

There are a couple of other choices disabling SELinux separation for the container

--security-opt label=disabled

Or creating a different policy type that can read (/write?) the content you want to volume mount in.

I am working on container_logreader_t right now to allow users to run a container that can read content in /var/log directory on the host.

**Comment by abitrolly** [comment]
How labels are stored? Is there only one SELinux label per one file?

**Comment by rhatdan** [comment]
Labels are stored as XAttrs on the file system.  There is only One label per file.

**Comment by rhatdan** [comment]
I am sure there is a limit on the length of a label, but it is large,  Serializing multiple labels into one string means nothing in SELinux.  There is one label per system object.  There is no concept of multiple labels.  And this is intentional.

**Comment by splatch** [comment]
Sorry for making this question, here but is there any useful piece of information over internet with description how to recover from mess caused by wrong mount suffixed with `:Z`? I managed to brick my non-selinux-enabled system boot because of binding `/lib` and `/etc` directories with this flag.
I checked with livecd the file attributes in /lib and /etc directories to see what context was set on them. So far I always get `?` in `ls -lZ` output. Non nonetheless I still can't chroot from livecd to my drive to fix the issue. The chroot command constantly returns `/bin/bash Permission denied`. By manual patch of chroot code I just confirmed that it gets errno 13 returned from [execvp call](https://github.com/coreutils/coreutils/blob/master/src/chroot.c#L426). I guess there might be a problem with system libraries or something which prevents program from loading, but I can't determine that.
Sadly I am unable to dig deeper into this issue. I hope there is still a way to revert what `:Z` did. Isn't it? (chcon doesn't work for me)

**Comment by rhatdan** [comment]
Are you talking about the hosts /lib and /etc or some other directory that is /lib and /etc on another system?

If they are the host /lib and /etc
restorecon -R -F /lib /etc
Should set the labels to the defaults.

I have no idea why a non selinux system would care if XATTRs are set on files and directories.
